docs/zkapps/advanced-snarkyjs.mdx

```diff
@@ -291,3 +291,170 @@ We imagine that one of the most important use cases for actions is to enable con
+
+</Alert>
+
+```ts
+class Leaderboard extends SmartContract {
+  // the root it the root hash of our off-chain Merkle tree
+  @state(Field) root = State<Field>();
+
+  // z is the hashed number we want to guess!
+  @state(Field) z = State<Field>();
+
+  deploy(args: DeployArgs) {
+    super.deploy(args);
+
+    // this is our hash we want to guess! its the hash of the preimage "22", but keep it a secret!
+    this.z.set(
+      Field(
+        "17057234437185175411792943285768571642343179330449434169483610110583519635705"
+      )
+    );
+  }
+
+  @method
+  guessPreimage(guess: Field, account: Account, path: MerkleWitness) {
+    // we fetch z from the chain
+    let z = this.z.get();
+    this.z.assertEquals(z);
+
+    // if our guess preimage hashes to our target, we won a point!
+    Poseidon.hash([guess]).assertEquals(z);
+
+    // we fetch the on-chain commitment
+    let commitment = this.commitment.get();
+    this.commitment.assertEquals(commitment);
+
+    // we check that the account is within the committed Merkle Tree
+    path.calculateRoot(account.hash()).assertEquals(commitment);
+
+    // we update the account and grant one point!
+    let newAccount = account.addPoints(1);
+
+    // we calculate the new Merkle Root, based on the account changes
+    let newCommitment = path.calculateRoot(newAccount.hash());
+
+    this.commitment.set(newCommitment);
+  }
+
+  @method
+  claimReward(account: Account, path: MerkleWitness) {
+    // we fetch the on-chain commitment
+    let commitment = this.commitment.get();
+    this.commitment.assertEquals(commitment);
+
+    // we check that the account is within the committed Merkle Tree
+    path.calculateRoot(account.hash()).assertEquals(commitment);
+
+    // we check that the account has at least 10 score points in order to claim the reward
+    account.score.assertGte(10);
+
+    // finally, we send the player a reward
+    this.send({
+      to: account.address,
+      100_000_000,
+    });
+  }
+}
+```
+
+Merkle trees allow us to easily reference off-chain data by only adding a couple of lines of code.
+However, it is important to mention that the developer of the zkApp must make sure that the Merkle tree that is being referenced on-chain is always in-sync with the actual off-chain data structure.
+
+You can look at [this example in the SnarkyJS repository](https://github.com/o1-labs/snarkyjs/tree/main/src/examples/zkapps/merkle_tree) to get a better understanding of how you can utilize the power of Merkle trees.
+
+### Merkle Tree - API reference
+
+```ts
+const treeHeight = 8;
+
+// creates a tree of height 8
+const Tree = new Experimental.MerkleTree(treeHeight);
+
+// creates the corresponding MerkleWitness class that is circuit-compatible
+class MerkleWitness extends Experimental.MerkleWitness(treeHeight) {}
+
+// sets a value at position 0n
+Tree.setLeaf(0n, Field.fromNumber(123));
```

Comment:
Is BigInt `0n` necessary or used to demonstrate what is considered best practice? Both are ok, just confirming.

Reply:
The API currently only accepts BigInts, the reason for that is if you "max out" your Merkle tree, the number range in JS won't be able to cover all leaves (2**(53-1) for `number` in JS, but we can have up to 2**(256-1) leaves, where 256 is the max height)